crates/astria-sequencer-relayer/src/relayer/celestia_client/mod.rs

```diff
@@ -0,0 +1,654 @@
+        let error = Error::BroadcastTxResponseErrorCode {
+            tx_hash: tx_response.txhash,
+            code: tx_response.code,
+            namespace: tx_response.codespace,
+            log: tx_response.raw_log,
+        };
+        return Err(error);
+    }
+    debug!(tx_hash = %tx_response.txhash, "broadcast transaction succeeded");
+    Ok(TxHash(tx_response.txhash))
+}
+
+/// Extracts the block height from the given response if available, or `None` if the transaction is
+/// not available yet.
+fn block_height_from_response(
+    response: Result<Response<GetTxResponse>, Status>,
+) -> Result<Option<i64>, Error> {
+    let ok_response = match response {
+        Ok(resp) => resp,
+        Err(status) => {
+            trace!(?status);
+            if status.code() == tonic::Code::NotFound {
+                debug!(msg = status.message(), "transaction still pending");
+                return Ok(None);
+            }
+            return Err(Error::FailedToGetTx(GrpcResponseError::from(status)));
+        }
+    };
+    let tx_response = ok_response
+        .into_inner()
+        .tx_response
+        .ok_or_else(|| Error::EmptyGetTxResponse)?;
+    if tx_response.code != 0 {
+        let error = Error::GetTxResponseErrorCode {
+            tx_hash: tx_response.txhash,
+            code: tx_response.code,
+            namespace: tx_response.codespace,
+            log: tx_response.raw_log,
+        };
+        return Err(error);
+    }
+    if tx_response.height == 0 {
+        debug!(tx_hash = %tx_response.txhash, "transaction still pending");
+        return Ok(None);
+    }
+
+    debug!(tx_hash = %tx_response.txhash, height = tx_response.height, "transaction succeeded");
+    Ok(Some(tx_response.height))
+}
+
+// Copied from https://github.com/celestiaorg/celestia-app/blob/v1.4.0/x/blob/types/payforblob.go#L174
+fn estimate_gas(blob_sizes: &[u32], cost_params: CelestiaCostParams) -> GasLimit {
+    // From https://github.com/celestiaorg/celestia-app/blob/v1.4.0/pkg/appconsts/global_consts.go#L28
+    const SHARE_SIZE: u64 = 512;
+    // From https://github.com/celestiaorg/celestia-app/blob/v1.4.0/pkg/appconsts/global_consts.go#L55
+    const CONTINUATION_COMPACT_SHARE_CONTENT_SIZE: u32 = 482;
+    // From https://github.com/celestiaorg/celestia-app/blob/v1.4.0/pkg/appconsts/global_consts.go#L59
+    const FIRST_SPARSE_SHARE_CONTENT_SIZE: u32 = 478;
+    // From https://github.com/celestiaorg/celestia-app/blob/v1.4.0/x/blob/types/payforblob.go#L40
+    const PFB_GAS_FIXED_COST: u64 = 75_000;
+    // From https://github.com/celestiaorg/celestia-app/blob/v1.4.0/x/blob/types/payforblob.go#L44
+    const BYTES_PER_BLOB_INFO: u64 = 70;
+
+    // From https://github.com/celestiaorg/celestia-app/blob/v1.4.0/pkg/shares/share_sequence.go#L126
+    //
+    // allow: we want pass-by-ref here to use it inside `Iterator::map`
+    #[allow(clippy::trivially_copy_pass_by_ref)]
+    fn sparse_shares_needed(blob_len: &u32) -> u64 {
```

Comment:
is `blob_len` the number of blobs or the total number of bytes in all the blobs?

Reply:
Neither :)  It's the size in bytes of one blob's `data` field.  I'll add a comment to clarify.

Reply:
Added in [c2a1e13](https://github.com/astriaorg/astria/pull/963/commits/c2a1e1393a7b594125d574e454d2f76bee6cfcac).